tests/api/runTx.js

```diff
@@ -0,0 +1,120 @@
+const { promisify } = require('util')
+const tape = require('tape')
+const Transaction = require('ethereumjs-tx')
+const runTx = require('../../lib/runTx')
+const StateManager = require('../../lib/stateManager')
+const { createAccount } = require('./utils')
+
+function setup () {
+  const vm = {
+    stateManager: new StateManager({ }),
+    emit: (e, val, cb) => { cb() },
+    runCall: (opts, cb) => cb(new Error('test'))
+  }
+
+  return {
+    vm,
+    runTx: promisify(runTx.bind(vm)),
+    putAccount: promisify(vm.stateManager.putAccount.bind(vm.stateManager)),
+    cacheFlush: promisify(vm.stateManager.cache.flush.bind(vm.stateManager.cache))
+  }
+}
+
+tape('runTx', (t) => {
+  const suite = setup()
+
+  t.test('should fail to run without opts', async (st) => {
+    shouldFail(st, suite.runTx(),
+      (e) => st.ok(e.message.includes('invalid input'), 'should fail with appropriate error')
+    )
+    st.end()
+  })
+
+  t.test('should fail to run without tx', async (st) => {
+    shouldFail(st, suite.runTx({}),
+      (e) => st.ok(e.message.includes('invalid input'), 'should fail with appropriate error')
+    )
+    st.end()
+  })
+
+  t.test('should fail to run without signature', async (st) => {
+    const tx = getTransaction()
+    shouldFail(st, suite.runTx({ tx }),
+      (e) => st.ok(e.message.toLowerCase().includes('signature'), 'should fail with appropriate error')
+    )
+    st.end()
+  })
+
+  t.test('should fail without sufficient funds', async (st) => {
+    const tx = getTransaction(true, true)
+    shouldFail(st, suite.runTx({ tx }),
+      (e) => st.ok(e.message.toLowerCase().includes('enough funds'), 'error should include "enough funds"')
+    )
+    st.end()
+  })
+})
+
+tape('should fail when runCall fails', async (t) => {
+  const suite = setup()
+
+  const tx = getTransaction(true, true)
+  const acc = createAccount()
+  await suite.putAccount(tx.from.toString('hex'), acc)
+  await suite.cacheFlush()
+
+  shouldFail(t,
+    suite.runTx({ tx, populateCache: true }),
+    (e) => t.equal(e.message, 'test', 'error should be equal to what the mock runCall returns')
+  )
+
+  t.end()
+})
```

Comment:
Nice cases. In addition to all the failure cases, can you add one which actually passes?

Reply:
Yeah, I think it is better to have this explicitly defined rather than have this implicitly tested somewhere else. Thanks for adding!